granary/flickr.py

```diff
@@ -0,0 +1,371 @@
+                                     owner.get('nsid')),
+        },
+      }
+
+    if isinstance(photo.get('tags'), dict):
+      activity['object']['tags'] = [{
+          'objectType': 'hashtag',
+          'id': self.tag_uri(tag.get('id')),
+          'url': 'https://www.flickr.com/search?tags={}'.format(
+            tag.get('_content')),
+          'displayName': tag.get('raw'),
+        } for tag in photo.get('tags', {}).get('tag', [])]
+    elif isinstance(photo.get('tags'), basestring):
+      activity['object']['tags'] = [{
+        'objectType': 'hashtag',
+        'url': 'https://www.flickr.com/search?tags={}'.format(
+          tag.strip()),
+        'displayName': tag.strip(),
+      } for tag in photo.get('tags').split(' ') if tag.strip()]
+
+    self.postprocess_activity(activity)
+    return activity
+
+  def comment_to_object(self, comment, photo_id):
+    """Convert a Flickr comment json object to an ActivityStreams comment.
+
+    Args:
+      comment: dict, the comment object from Flickr
+      photo_id: string, the Flickr ID of the photo that this comment belongs to
+
+    Returns:
+      dict, an ActivityStreams object
+    """
+    obj = {
+      'objectType': 'comment',
+      'url': comment.get('permalink'),
+      'id': self.tag_uri(comment.get('id')),
+      'inReplyTo': [{'id': self.tag_uri(photo_id)}],
+      'content': comment.get('_content', ''),
+      'created': self.reformat_unix_time(comment.get('datecreate')),
+      'updated': self.reformat_unix_time(comment.get('datecreate')),
+      'author': {
+        'objectType': 'person',
+        'displayName': comment.get('realname'),
+        'username': comment.get('authorname'),
+        'id': self.tag_uri(comment.get('author')),
+        'url': self.user_url(comment.get('path_alias')),
+        'image': {
+          'url': self.get_user_image(comment.get('iconfarm'),
+                                     comment.get('iconserver'),
+                                     comment.get('author')),
+        },
+      }
+    }
+    self.postprocess_object(obj)
+    return obj
+
+  def get_user_image(self, farm, server, author):
+    """Convert fields from a typical Flickr response into the buddy icon
+    URL.
+
+    ref: https://www.flickr.com/services/api/misc.buddyicons.html
+    """
+    return 'https://farm{}.staticflickr.com/{}/buddyicons/{}.jpg'.format(
+      farm, server, author)
+
+  def reformat_unix_time(self, ts):
+    """Convert a Unix timestamp into a datetime object
+
+    Args:
+      ts: string or int, the unix timestamp
+
+    Returns:
+      a datetime.datetime
+    """
+    return ts and datetime.datetime.fromtimestamp(int(ts)).isoformat()
+
+  def user_url(self, handle):
+    """Convert a user's screen name to their Flickr profile page URL.
+
+    Args:
+      handle: string, the Flickr user's screen name
+
+    Returns:
+      string, a profile URL
+    """
+    return handle and 'https://www.flickr.com/people/%s' % handle
```

Comment:
does this url format also work for user ids? i think that's what `user_url()` will at least sometimes be given.